Revamped UCLA Radio Android

arjunr00 opened the pull request:
## Overview
Visually, pretty much gave the app a whole new makeover: added a background, some layout changes, and gave it a splash screen, among other things. I also changed up the app icon; I like this one a little better but the old one's still in there if needed.
Functionality-wise, I moved the streaming code into its own Service so that it could a) load and run on its own, separate from the UI thread and the main activity's lifecycle and b) be controlled while the user is using other apps/the device is locked. Good thing was, all the REST boilerplate stuff with Retrofit, including functions and interfaces, was already implemented so that saved me the trouble.
More details and specifics can be found in the commit history.

## Types of changes

- [x] Bugfix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [x] Refactor (change which changes the codebase without affecting its external behavior)
- [x] Non-breaking change (fix or feature that would causes existing functionality to work as expected)
- [ ] Breaking change (fix or feature that would cause existing functionality to __not__ work as expected)

## Purpose

We need an Android app lol

## Screenshot(s) 

**Video [here](https://photos.app.goo.gl/L6lj4Wlu2DlXlE0h2).**
Click on each image below to enlarge.

| Splash Screen | Stream | Shows | DJs | About | Notification |
|---|---|---|---|---|---|
|![image](https://user-images.githubusercontent.com/30734384/39962269-714eba88-55fe-11e8-89a6-695cf4e704cc.png)|![image](https://user-images.githubusercontent.com/30734384/39962672-10f410e0-560b-11e8-8304-3c1f7d4ea0bb.png)|![image](https://user-images.githubusercontent.com/30734384/39962271-783142d0-55fe-11e8-935b-7cdb09896135.png)|![image](https://user-images.githubusercontent.com/30734384/39962272-7a3916b6-55fe-11e8-91fc-8a3927c7d176.png)|![image](https://user-images.githubusercontent.com/30734384/39962273-7c751056-55fe-11e8-9798-24a39f27d176.png)|![image](https://user-images.githubusercontent.com/30734384/39962677-1ca50750-560b-11e8-934c-0259bbd1da5c.png)|

## Known Issues

- When the user pauses the stream and resumes, it picks up from where it left off instead of keeping synced with the actual station in real time. It turns out that (I think) this is a problem with SHOUTcast and not the app, because the same effect happens on the website and at uclaradio.com:8000.
- If the connection to the server is dropped for too long (I haven't figured out the exact time, but it's definitely only for a while, maybe a few minutes?), the app won't reconnect; the user'll have to restart it.
- On the DJs grid, probably due to the CircleImageView library, the images pop into the view abruptly even after Picasso loads them.

## Further Comments

As of now the only device this app has really been tested on is my own, which is a few years old now, running Android 7.1.1 (Nougat).

arjunr00 committed:
Update StreamingFragment to include show art and title

Minor changes/fixes:
  - Updated Picasso to most recent version
  - Added visual toggling for play/pause button
  - Added function to get currently-playing show to RadioPlatform
  - "Fixed" app/build.gradle (more like wrapped it in duct tape
    temporarily while waiting for an actual fix)

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Schedule now separates shows based on the day they're on air

More detail: Added an extra TextView to the schedule_item layout, then
controlled its visibility based on whether the current item's day is
different from that of the previous item.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Added functionality to call On-Air or Requests via two buttons on the
streaming fragment

More details: Updated permissions to include CALL_PHONE; app requests
permission on button press (for versions >=6.0)

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Fixed permissions check callback

Also removed some unused variables from StreamingFragment.java

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Stream now loads behind a (placeholder) splash screen

More details: The MediaPlayer now runs on a Service, and prepares behind
a splash screen so that the stream can be immediately started on the
main screen. This is still quite rough; there are some lifecycle issues
with the Service/MediaPlayer (the service appears to restart upon exit,
for example). Requires more testing and polish, but the general func-
tionality is there.

Minor changes:
  - Renamed package, need to rename others to follow convention

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Fixed service lifecycle issues and added persistent notification

More details: Now the StreamService starts in the foreground to avoid
being killed by the OS for memory. For this it now has a placeholder
notification that will be turned into controls for the MediaPlayer.
Upon destruction the service unbinds and kills itself as required.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Made splash screen a little nicer

More details: Also made sure that the service stops in case the user
exits out of the app on the splash screen while the MediaPlayer is
preparing (otherwise there'd be a rogue service running in the back-
ground eating resources and battery and we wouldn't want that).

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Added "pre-splash" screen for seamless start-up

More details: The main splash screen has to inflate a layout, start a
service, and listen for a Broadcast, so there was a little lag on
startup--users would see a blank white screen BEFORE the splash screen,
which is unacceptable. Now, there's a pre-splash screen with only the
logo displayed that starts up almost immediately and then hands off to
the main splash screen when the latter's ready. (For now, this pre-
splash screen should only work for devices on Android 6.0+ because of
compatibility issues, but I still need to test that older versions jump
straight to the main splash screen as required).

NOTE: Errors are being reported in app/res/values/styles.xml but they
don't seem to be impacting the app or causing build errors in any way,
so I'm just leaving them alone.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Moved notification code into StreamService

Minor changes include cleaning up imports in MainActivity

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Notification now displays show art and title; info changes on the hour

More details: Notification and StreamingFragment sync up playing/
pausing. Every 10 seconds, the app checks the time; if it's a new hour,
it tries to update the show art and title. Needs more rigorous testing.

Minor fixes: ViewPager now keeps all four tabs in memory, preventing
loss of show art/title while playing. More unused import removal.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Minor commit: increase the times StreamService looks for new data

More details: Before, it checked every 10 seconds for a minute. Now,
it checks every 20 seconds for 3 minutes. (Before: 6 checks in a shorter
interval; Now: 9 checks in a longer interval). This gives the app a
lower chance of missing an update.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Minor commit: Changed default showTitle to "Loading show..."

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Minor visual overhaul: background, app colors... See more details.

More details: Added a background to all activities. Action bar, status
bar, and nav bar now have a background of #80333333 (#333333 at 50%
opacity). Only small changes, but they go a long way in making the app
look nicer.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Added some basic stream error handling

More details: Now, on startup, if the app is connected to a WiFI network
that blocks the Radio stream (right now this list only includes UCLA_
WEB and UCLA_WEB_RES), it will prompt the user to switch networks. Also,
if the stream drops connection temporarily, the app will now attempt to
reconnect and continue (before, it would simply treat the end of the
stream as the end of the MediaPlayer's input and loop whatever was
recorded).

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Bugfix: Catches a possible IllegalStateException from isPlaying()

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Minor update to splash screen - navbar backgound is transparent

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Added new launcher icons; changed notification icon; fixed some bugs.

More details: New notification icon is just the favicon for the website.
Added a color (trianglified) version of the old launcher icon as well as
a circle bear icon (for fun I guess?).

Bug fixes: If the DJ or Schedule list fail to update for some reason,
they will now retry instead of just leaving a blank Fragment. Also,
SplashActivity was stopping StreamService whenever it was destroyed;
this was a problem if the user stopped the activity while it was
loading, because then the MediaPlayer was killed and not revived,
leading to a NullPointerException when isPlaying() was called. This has
been fixed.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Minor fix: prevented splash screens from changing screen orientation.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Added color to notification based on vibrant color of show art

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Added bottom sheet for shows (tapping a show brings up more details)

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Fixed show art in update method so it resets to the bear.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Minor changes/fixes, see more details below.

More details:
  - Bottom sheets for show details now displays the bear in the case
    of a missing image
  - Moved map->string parsing for DJ JSON data into ScheduleData
  - Added progress loading circles for streaming show art as well as
    the shows and DJs fragments so the user isn't staring at a blank
    screen while they load.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Fixed bug with stream when connection drops and comes back.

More details: Before, the stream was lagging behind for some unknown
reason (perhaps it was still looping?), so now it properly resets the
stream. However, the user must now manually hit the play button in
order to continue the stream.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Changed Shows fragment UI to a 2-column grid

More details: Just think this looks a little nicer; I've left the old
layout file/code just in case people prefer the old style. Also: missing
show art now defaults to the same art as on the website, which looks
better anyway, and logo_banner_white.png was shrunk by 50% because it
was so large that there was an occasional OOM error thrown by the app on
start-up.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Made play/pause button a little nicer-looking (it's a circle now)

More details: Truthfully, the circle button as it is right now looks
a little out of place. That's mainly because it's in preparation for
a revamp of the streaming UI, and it doesn't look TOO bad anyway. Also,
it changes color with the show art as well.

Minor additions: Actual play/pause icons were added (obtained from
material.io/tools/icons) because the default ones are ugly as hell.

Minor fixes: OOM errors were still cropping up occasionally, so now
if the app can't load the banner logo on the UI thread it'll delegate
it to Picasso to load it aynchronously. This will create some visual
lag, but it happens rarely enough that it shouldn't be too bad an issue.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Moved hardcoded strings into string resource file /res/values/strings.xml

Minor fixes: Fixed warnings such as unused imports, possible NPEs, etc.
Also switched all layouts to ConstraintLayouts if they weren't before
(which are a lot more responsive and flexible with different screen
sizes. Fixed a small bug with the color of the play button.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Fixed handling of connection drops; stream resumes automatically.

More details: In addition, there's a visual indication of a stream drop
now so that the user doesn't think something's wrong with the app or
the server. Also, if the user pauses the stream and then hits play, it
resumes from the point where they hit pause, and NOT the actual current
position of the stream instead. After quite a bit of grueling work, upon
testing against the actual website it turns out this is because of the
way SHOUTcast works, and has nothing to do with the app. Unfortunately,
this means that if the user has the app open and paused for a while it
will run quite out of sync with the server unless they restart it.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Added debug build type and set up release to be more efficient.

Also removed the trianglify library as it's not being used.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Reorganized and renamed packages to follow convention.

Also removed unused/unnecessary classes.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

comran commented on app/src/main/java/com/uclaradio/uclaradio/activities/MainActivity.java:
```diff
@@ -0,0 +1,147 @@
+        DJsFragment.OnFragmentInteractionListener,
+        AboutFragment.OnFragmentInteractionListener {
+
+  public static final int SERVICE_ID = 41243;
```
Should be an XML resource

arjunr00 replied:
Done.

comran commented on app/src/main/java/com/uclaradio/uclaradio/activities/MainActivity.java:
```diff
@@ -0,0 +1,147 @@
+    BitmapDrawable background_drawable = new BitmapDrawable(Bitmap.createScaledBitmap(background, width * 4, actionBarHeight * 4, false));
+    background_drawable.setTileModeX(android.graphics.Shader.TileMode.REPEAT);
+//    actionBar.setBackgroundDrawable(background_drawable);
+    int color = Color.parseColor("#80333333");
```
Custom color goes in values/colors.xml resource file

arjunr00 replied:
Done.

comran commented on app/src/main/java/com/uclaradio/uclaradio/activities/SplashActivity.java:
```diff
@@ -0,0 +1,91 @@
+    //  it will be more efficient and flexible to determine the block automatically.
+    public static ArrayList<String> badWifis
+            = new ArrayList<>(Arrays.asList(
+            "\"UCLA_WEB\"", "\"UCLA_WEB_RES\""
```
Resource file

Also -- the stream doesn't work on these wifis because UCLA_WEB_* blocks all ports except for http/https. Other schools do this too, so it may be worth trying to do some sort of ping to a non-http/https server and see if the local network is blocking ports

arjunr00 replied:
Moved into resource, will work on a ping thing soon.

arjunr00 replied:
Done.

comran commented on app/src/main/java/com/uclaradio/uclaradio/fragments/about/AboutFragment.java:
```diff
@@ -74,7 +73,7 @@ public void onDetach() {
+        ImageButton fbk = view.findViewById(R.id.button1);
         fbk.setOnClickListener(new View.OnClickListener() {
             public void onClick(View v) {
                 Uri uri = Uri.parse("https://www.facebook.com/UCLARadio/?fref=ts");
```
should also be in resource xml (string)

comran replied:
along with the other stuff on the about page

arjunr00 replied:
Done.

comran commented on app/src/main/java/com/uclaradio/uclaradio/fragments/djs/DJsFragment.java:
```diff
@@ -31,17 +32,21 @@
+    private final int numberOfCols = 2;
+
+    // These aren't being used anywhere, but I don't want to delete some of the boilerplate
+    //  in case it comes in handy later
```
Deleting parameter stuff for now would clean up the code a fair amount though -- we always have git to look back at old code in case we do want to send parameters

arjunr00 replied:
Done.

comran commented on app/src/main/java/com/uclaradio/uclaradio/fragments/djs/DJsFragment.java:
```diff
@@ -110,45 +115,53 @@ public void onDetach() {
-                .baseUrl("https://uclaradio.com/")
-                .build();
+        .addConverterFactory(GsonConverterFactory.create())
+        .baseUrl("https://uclaradio.com/")
```
string resource

arjunr00 replied:
Done.

comran commented on app/src/main/java/com/uclaradio/uclaradio/fragments/djs/DJsFragment.java:
```diff
@@ -110,45 +115,53 @@ public void onDetach() {
                     @Override
                     public void onFailure(Call<DjList> call, Throwable t) {
                         Log.e("TAG", "FAILED TO MAKE API CALL");
+                        getDjs();
```
this could infinite loop w/o any wait time -- maybe add a sleep so that it polls at a slower rate, in the worse case?

arjunr00 replied:
Now runs in a delayed Handler whose delay doubles each time it's run.

comran commented on app/src/main/java/com/uclaradio/uclaradio/fragments/schedule/ScheduleAdapter.java:
```diff
@@ -0,0 +1,280 @@
+    resources = context.getResources();
+    baseUrl = resources.getString(R.string.website);
+
+    dayToNum.put("Sun",0); dayToLongDay.put("Sun", resources.getStringArray(R.array.days_of_week)[0]);
```
this block of code could be simplified to use an array list of "sun", mon", "tue", ... and a for loop

arjunr00 replied:
Done.

comran commented on app/src/main/java/com/uclaradio/uclaradio/fragments/schedule/ScheduleAdapter.java:
```diff
@@ -0,0 +1,280 @@
+    for (int i = 0; i < items.size(); i++) {
+      ScheduleData item = items.get(i);
+      // Some trickery going on here
+      if (i % 2 == 1 && (items.get(i-1) != null && !item.getDay().equals(items.get(i-1).getDay()))) {
```
Add a comment for what exactly is going here

arjunr00 replied:
Done.

comran commented on app/src/main/java/com/uclaradio/uclaradio/fragments/schedule/ScheduleAdapter.java:
```diff
@@ -0,0 +1,280 @@
+            });
+  }
+
+  //region Old onBindViewHolder
```
Is this commented out code still needed in the near future?

arjunr00 replied:
I don't see it being needed, so removed.

comran commented on app/src/main/java/com/uclaradio/uclaradio/fragments/schedule/ScheduleFragment.java:
```diff
@@ -30,16 +32,18 @@
-    // TODO: Rename and change types of parameters
+    private ContentLoadingProgressBar scheduleProgress;
+
+    // Again, don't want to remove boilerplate so leaving this stuff here
```
same comment as above ^^

arjunr00 replied:
Done.

comran commented on app/src/main/java/com/uclaradio/uclaradio/fragments/streaming/StreamingFragment.java:
```diff
@@ -0,0 +1,308 @@
+
+  private final Intent callIntent = new Intent(Intent.ACTION_CALL);
+
+  private static final int REQUEST_PHONE_CALL = 982;
```
resource xml

arjunr00 replied:
Done.

comran commented on app/src/main/java/com/uclaradio/uclaradio/fragments/streaming/StreamingFragment.java:
```diff
@@ -0,0 +1,308 @@
+          return;
+        }
+        if (MainActivity.stream == null) {
+          Log.d("Service", "Stream is null...");
```
Give the user an alert -- a null stream would mean that the stream controls do nothing

arjunr00 replied:
Done.

comran commented on app/src/main/java/com/uclaradio/uclaradio/fragments/streaming/StreamingFragment.java:
```diff
@@ -0,0 +1,308 @@
+    onAirCallBtn.setOnClickListener(new View.OnClickListener() {
+      @Override
+      public void onClick(View v) {
+          callIntent.setData(Uri.parse("tel:3107949348"));
```
resource

arjunr00 replied:
Done.

comran commented on app/src/main/java/com/uclaradio/uclaradio/fragments/streaming/StreamingFragment.java:
```diff
@@ -0,0 +1,308 @@
+    requestCallBtn.setOnClickListener(new View.OnClickListener() {
+      @Override
+      public void onClick(View v) {
+        callIntent.setData(Uri.parse("tel:3108259999"));
```
^

arjunr00 replied:
Done.

comran commented on app/src/main/java/com/uclaradio/uclaradio/fragments/streaming/StreamingFragment.java:
```diff
@@ -0,0 +1,308 @@
+                              .generate();
+
+                      int newBgColor = color.getVibrantColor(value.data);
+                      int newFgColor = color.getDarkVibrantColor(Color.parseColor("#FFFFFF"));
```
@color/white

arjunr00 replied:
Done.

comran commented on app/src/main/java/com/uclaradio/uclaradio/fragments/streaming/StreamingFragment.java:
```diff
@@ -0,0 +1,308 @@
+      Palette color = Palette
+              .from(((BitmapDrawable) showArtIv.getDrawable()).getBitmap())
+              .generate();
+      int newFgColor = color.getDarkVibrantColor(Color.parseColor("#FFFFFF"));
```
^

arjunr00 replied:
Done.

comran commented on app/src/main/java/com/uclaradio/uclaradio/fragments/streaming/StreamingFragment.java:
```diff
@@ -0,0 +1,308 @@
+          Palette color = Palette
+                  .from(((BitmapDrawable) showArtIv.getDrawable()).getBitmap())
+                  .generate();
+          int newFgColor = color.getDarkVibrantColor(Color.parseColor("#FFFFFF"));
```
^

arjunr00 replied:
Done.

comran commented on app/src/main/java/com/uclaradio/uclaradio/stream/StreamService.java:
```diff
@@ -0,0 +1,367 @@
+public class StreamService extends Service implements MediaPlayer.OnPreparedListener {
+    private static final String STREAM_URL = "http://uclaradio.com:8000/;";
+    public  static final String BROADCAST_ACTION = "com.uclaradio.uclaradio.BROADCAST_COMPLETE";
+    private static final String CHANNEL_ID = "com.uclaradio.uclaradio.notificationChannelId124813759";
```
all should be string resources (xml)

arjunr00 replied:
Done.

comran commented on app/src/main/java/com/uclaradio/uclaradio/stream/StreamService.java:
```diff
@@ -0,0 +1,367 @@
+//                sendBroadcast(new Intent(getString(R.string.play_pause_intent)));
+            }
+        });
+        stream.setOnErrorListener(new MediaPlayer.OnErrorListener() {
```
Report stream errors to the user as well

arjunr00 replied:
Done.

comran reviewed (commented):
Overall actually looks pretty good -- these are mainly stylistic stuff + cleanup that should happen on the android code to make it more maintainable in the future too

I checked out this branch to check out, but my android studio has a ton of updates rn -- will post another update in the future if it runs successfully/has issues

arjunr00 committed:
Applied requested changes: Resources, remove unused/commented code

More details: Moved remaining constant strings, integers, and colors to
resource .xml files (strings.xml, integers.xml, and colors.xml, res-
pectively. Also removed large blocks of unused/commented code.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Deleted file that I made but never used and forgot to delete

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Applied requested changes: report stream errors to user.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Applied requested changes: SplashActivity now tries to ping the server.

More details: Rather than checking the SSID against known blockers, the
activity now tries to ping the server with a 10 second timeout, and
displays a notice accordingly.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 commented:
All requested changes have been addressed!

comran commented:
Finally managed to get Android studio gradle to sync (didn't have good internet before)

Stream + error handling when on a port blocking wifi works well now. I'm getting pretty noticeable slowdowns when swiping between tabs, and logcat is reporting a bunch of these types of errors:

> 05-15 00:52:13.878 13848-13935/com.uclaradio.uclaradio E/OpenGLRenderer: com.uclaradio.uclaradio/com.uclaradio.uclaradio.activities.MainActivity continue jank frame too much
05-15 00:52:15.188 13848-13935/com.uclaradio.uclaradio E/OpenGLRenderer: com.uclaradio.uclaradio/com.uclaradio.uclaradio.activities.MainActivity continue jank frame too much
05-15 00:52:16.068 13848-13935/com.uclaradio.uclaradio E/OpenGLRenderer: com.uclaradio.uclaradio/com.uclaradio.uclaradio.activities.MainActivity continue jank frame too much
05-15 00:52:16.988 13848-13935/com.uclaradio.uclaradio E/OpenGLRenderer: com.uclaradio.uclaradio/com.uclaradio.uclaradio.activities.MainActivity continue jank frame too much

This is indicating that the app is performing too much other work on its Main thread, which will cause noticeable UI slowdown. I haven't dug deep into the root cause of the problem, but the profiler may be of help for fixing the slowdown issues. Let me know if these errors are not showing up for you (I ran it on my phone, so the issues should be apparent on other devices as well)

comran commented:
Also the stream continues to use data even after being paused. The original code (from what I remember) got around this issue by stopping the stream after every pause. A better system may involve stopping after a set timeout when in the app or on a user state event (like exiting the app or screen off).

From exiting the UCLA radio app after playing the stream and then pausing, it looks like it is still using 4G data. If the user of the app is using data for the stream then this could cause high quota usages and bad battery life even when the user thinks the stream was paused

arjunr00 commented:
Putting in a timeout is a good idea; maybe after 15-30 seconds of being paused the MediaPlayer can stop, and if the user hits play after that the app should just notify them that it's loading the stream (probably through a Toast or something). This won't be hard to implement, so I'll fix this up asap.

As for the slowdown when swiping between tabs you're experiencing: I'm not seeing any lag when swiping between tabs on my own phone, and my LogCat isn't saying anything about dropped frames either (neither are those specific errors showing up on my device). Could you give me more details about the conditions under which it's lagging? i.e. is it only happening when you're on 4G and not WiFi, perhaps, or if you have more apps running in the background, etc. 

Looking at the memory profiler:

![image](https://user-images.githubusercontent.com/30734384/40074831-a3e067c0-582f-11e8-8dfe-ee220bf63d74.png)

It's maintaining a pretty steady level of usage just under 256MB (the fluctuations pop up when I scroll through the DJs or the shows. I do think that I can bring the graphics memory usage down, though: there are some drawables with just one size that are in res/drawables; these should have multiple sizes that go into specific drawable folders (the background is 900x1440 px, which is overkill for lower-dpi devices). Truthfully I'm not entirely sure if this memory usage is what's causing slowdowns for you, but this is anyway a good change to make so I'll get on this as well.

arjunr00 committed:
Pausing for a certain amount of time now stops the stream.

More details: This way, the stream isn't constantly using up data while
paused; after a certain amount of time (defined in integers.xml), the
stream will stop, and if the user then hits play the icon will change
to a reloading icon, notifying the user that the stream is loading.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 committed:
Split all drawables into dpi-specific files; removed unused drawables.

More details: This has had a drastic effect on memory usage, almost
halving it! This was a gross oversight that should have been corrected
far earlier. Also: some code in initializeActionBar() was not being
used, so that's been removed.

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

arjunr00 commented:
Sure enough, splitting up those drawables into dpi-specific versions cut down memory usage quite a bit indeed, it's sitting around the 150 MB mark:

![image](https://user-images.githubusercontent.com/30734384/40080569-35a09d5a-5840-11e8-8de0-a3334869c4b0.png)

If your phone has a high pixel density, then this would explain why it was slowing down more for you while remaining pretty smooth for me--my phone falls under the hdpi category, so the OS wasn't scaling the drawables up as much. (Although you did say that there was lag when switching between tabs specifcally. The fluctuation in memory usage when swiping between tabs is just around 5 MB, at least on my device, but while I don't _think_ there's something else going on the background eating memory or causing lag, I could still be wrong).

Let me know if it's smoother now! (If it's not, could you give me a little more information on your device and the conditions under which it's lagging?)

arjunr00 committed:
Minor commit: Changed duration of connection error Toasts

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

comran commented:
LGTM -- slowdown issues are fixed + stream stops using data after a certain timeout after a pause event. One minor thing that remains is a glitch on the startup screen ucla radio logo that I'm not sure about the cause for (it's on the top left of the logo)

![screenshot_2018-05-15-16-06-31](https://user-images.githubusercontent.com/19298617/40088110-6e713c4c-585a-11e8-9a5a-a9f2c2f132c5.jpg)

But otherwise, should be good to go

arjunr00 commented:
Oh whoops, yeah my copy of Photoshop has had a weird bug recently where occasionally it'll introduce these weird bars into random spots in the image that won't go away without reverting to the last time I saved--never figured out how to fix that. I'll just fix that one up and make sure the other images are fine.

arjunr00 committed:
Fixed bugged-out images (and removed remaining unused drawables)

Signed-off-by: Arjun Raghavan <arjun.raghavan.2k@gmail.com>

comran reviewed (approved):
Yep that fixed the issue
